docs/tm/udq_actionx/udq.tex

```diff
@@ -0,0 +1,596 @@
+\inlinecode{SummaryState} instance\footnote{Observe that during initialization
+the \udq{} expressions are inspected, and we make sure that all summary
+variables needed to evaulate \udq{} expressions are evaluated in the Summary
+evaluation.}. The \inlinecode{SummaryState} class manages a set of maps with
+well, group and field variables, when evaluating e.g. the \kw{WOPR} the results
+will be stored in a two level map first indexed with keyword \kw{WOPR} and then
+with well name. Afterwards the \udq{} layer can fetch values with
+\inlinecode{SummaryState::get\_well\_var()}. The values in the
+\inlinecode{SummaryState} have been converted to output units, this is important
+for the \kw{UDA} evaluation.
+
+At the end of every timestep the \inlinecode{UDQConfig::eval()} method is called
+to evaluate all the \udq{} expressions, the evaluated values will end up in the
+active \inlinecode{SummaryState} instance, i.e. for this \udq{}
+\begin{deck}
+UDQ
+  DEFINE FUOPR SUM(WOPR) /
+  DEFINE WUGWR WGPR / WWPR /
+/
+\end{deck}
+we will get \inlinecode{SummaryState} entries for \kw{FUOPR} and \kw{WUGWR} for
+\emph{all} wells in the model. These \inlinecode{SummaryState} values can then
+be output to the summary file, be used when evaluating \kw{ACTIONX} keywords or
+to evaluate \kw{UDA} control values.
+
+In addition to the \inlinecode{SummaryState} variable there is an instance of
+type \inlinecode{UDQState} which is updated runtime, the \inlinecode{UDQState}
+instance holds on to the results of \udq{} evaluations with more context than
+\inlinecode{SummaryState} and is used to output \udq{} and \kw{UDA} state to the
+restart files. While evaluating the simulator will create a \kw{UDQContext}
+variable which will manage both the \inlinecode{SummaryState},
+\inlinecode{UDQState} and also some \udq{} parameters from the \kw{RUNSPEC}
+section. The lifetime of the \kw{UDQContext} instance is only one \udq{}
+evaluation.
+
+While evaluating the \udq{} expressions the results will be instances of
+\inlinecode{UDQSet} which is a small container class which keeps track of
+well/group names and whether a value is defined or not. The \inlinecode{UDQSet}
+class overrides the arithmetic operators like \inlinecode{UDQSet::operator+()}
+so that expressions like \kw{2 * WOPR 'OP*'} can be easily evaluated in code.
+
+\subsection*{Paralell awareness}
+\label{udq_parallel}
+The opm-common code where the \udq{} functionality is implemented is totally
+unaware of parallel execution, so to be certain that this works for a parallel
+simulator care must be taken. In flow this is handled as:
+\begin{enumerate}
+  \item The \inlinecode{Schedule} class as a whole is identical on all
+    processes.
+  \item The state variables \inlinecode{UDQState} and \inlinecode{SummaryState}
+    are distributed so they are equal on all processes before the \udq{}
+    evaluation. This communication is performed in the simulator.
```

Comment:
Done nicely for the SummaryState :clap: 
I could find where this is done for SummaryState (in [GenericEclWriter::evalSummary)](https://github.com/OPM/opm-simulators/blob/master/ebos/eclgenericwriter.cc#L513-L523)), but what I am missing is where this is done for UDQState.
Or is that done somewhere by reconstructing it from the distributed SummaryState?
Or is UDQState only needed on the output process?

Reply:
> Or is UDQState only needed on the output process?

The `UDQState` is only read and updated in the opm-common code, i.e. although it is *owned* by the simulator the simulator does *not* update it - only pass to to `UDQConfig::eval()` as an opaque handle. Therefor all access is trivially parallel correct (if there is such a thing!)